AMCL : : External pose fusion

dimaxano opened the pull request:
## Purpose
We can improve localization by fusion external pose source (which we trust to some extent) to the AMCL.

## Approach

Implementation is taken from [here](https://github.com/midemig/gps_amcl).
What in the essence have changed:
1. We reassign weights of particles: the closer particle pose estimate to external pose the higher the weight, roughly speaking
2. Added random particle generation (with some probability) around current external pose.

## Dependencies
 - cmrobotics/cmr_elf#5
 
## Learning
- Paper which fusion algorithm is based on: [Improved LiDAR Probabilistic Localization for Autonomous Vehicles Using GNSS](https://www.mdpi.com/1424-8220/20/11/3145)

dimaxano committed:
add fusion logic, WIP

dimaxano committed:
initial working commit

dimaxano committed:
remove redundant printing

cmraaron commented on nav2_amcl/src/amcl_node.cpp:
```diff
@@ -562,6 +588,61 @@ AmclNode::initialPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::Sha
+      msg.pose.pose.orientation.w);
+  tf2::Matrix3x3 m(q);
+  double roll, pitch, yaw;
+  m.getRPY(roll, pitch, yaw);
```
looks like you are only using `yaw` here?  I believe this (Quarternion, Matrix, extra locals) can be replaced with `tf2::getYaw(msg.pose.pose.orientation)`

dimaxano replied:
Yep, makes sense!

cmraaron commented on nav2_amcl/include/nav2_amcl/pf/pf.hpp:
```diff
@@ -131,17 +131,22 @@ typedef struct _pf_t
+
+  double gps_x, gps_y, gps_yaw;
+  double cov_matrix[9];
+  double eigen_matrix[9];
```
Would it make sense to use the c++ typesafe `Array<double, 9>` here?

dimaxano replied:
As we use that arrays inside C code, we do have an other option than just a simple array

cmraaron commented on nav2_amcl/include/nav2_amcl/pf/pf.hpp:
```diff
@@ -131,17 +131,22 @@ typedef struct _pf_t
+  double dist_threshold; //distance threshold in each axis over which the pf is considered to not be converged
+  int converged; 
+
+  double gps_x, gps_y, gps_yaw;
```
these are not strictly gps now right?  Give these are only valid as a group, does it make sense to put these in their own struct?

dimaxano replied:
As for separate data structure question, I will use the right given by you to resolve comment without changes :)
Do not want to introduce new data structure here, but `gps` prefix is worth removing, ofc

cmraaron commented on nav2_amcl/include/nav2_amcl/sensors/external_pose/external_pose.hpp:
```diff
@@ -0,0 +1,60 @@
+    double eigen_matrix[9];
+
+    double time_ns; // time the measurement was made
+} external_pose_t;
```
could use the c++ `struct ExternalPose` without the `typedef`

dimaxano replied:
Agree

cmraaron commented on nav2_amcl/src/amcl_node.cpp:
```diff
@@ -393,6 +399,26 @@ AmclNode::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+  if (d.nanoseconds() * 1e-9 > ext_pose_check_interval_.count()) {
+    RCLCPP_WARN(
+      get_logger(), "No external pose received for %f"
+      " seconds. Swithing it to inactive state", d.nanoseconds() * 1e-9);
```
just `d.seconds()`?, then wouldn't need the multiplication

dimaxano replied:
Yep

cmraaron commented on nav2_amcl/src/amcl_node.cpp:
```diff
@@ -562,6 +588,61 @@ AmclNode::initialPoseReceived(geometry_msgs::msg::PoseWithCovarianceStamped::Sha
+  double roll, pitch, yaw;
+  m.getRPY(roll, pitch, yaw);
+  double cov_matrix[9] = {msg.pose.covariance[0], msg.pose.covariance[1] ,msg.pose.covariance[5] ,msg.pose.covariance[6] ,msg.pose.covariance[7] ,msg.pose.covariance[11] ,msg.pose.covariance[30] ,msg.pose.covariance[31] ,msg.pose.covariance[35] };
+  memcpy(pose.cov_matrix, cov_matrix, 9*sizeof(double));
```
can just assign direct to `pose.cov_matrix`, it would be nine assignments, but perhaps not if its `Array<double, 9`.  Then wouldnt need to assemble on stack and then memcpy.

dimaxano replied:
I tried to dive deeper into memcpy vs asignment question, and it seems memcpy is not that bad in most of the cases with arrays of simple types

Attaching discussion on that
https://groups.google.com/g/comp.lang.c++/c/HWbSJacwGnU

cmraaron commented on nav2_amcl/src/sensors/external_pose/external_pose.cpp:
```diff
@@ -0,0 +1,37 @@
+void
+ExternalPoseBuffer::addMeasurement(const external_pose_t measurement)
+{
+    if(buffer_.size() >= max_buff_size_) buffer_.erase(buffer_.begin());
```
this will be `max_buff_size_` reallocations as its deleting from the start of a `vector`.   `std::queue` would be your friend here.  then would be `buffer_.pop_front(); buffer_.push_back();`

dimaxano replied:
But I need to have an element-wise access to the `buffer_` data structure, which is not provided for `std::queue`, right?

Maybe using heap for the `buffer_` is the best way to go here?

cmraaron commented on nav2_amcl/src/sensors/external_pose/external_pose.cpp:
```diff
@@ -0,0 +1,37 @@
+    }
+
+    if(min_abs_diff < search_tolerance_ns_){
+        out_measurement = buffer_.at(min_diff_idx);
```
`min_diff_idx` is never assigned.  I assume it should be done along with `min_abs_diff`

dimaxano replied:
Huh, what a bug
Thank for catching it!

cmraaron commented on nav2_amcl/src/pf/pf.c:
```diff
@@ -289,6 +293,77 @@ void pf_update_sensor(pf_t * pf, pf_sensor_model_fn_t sensor_fn, void * sensor_d
+    pose[1] = pose[1] + gps_y;
+    pose[2] = pose[2] + gps_yaw;
+
+    memcpy(pose_v, pose, 3*sizeof(double));
```
just assign directly into `pose_v`?  also do you want to remove the `gps_` prefix?

dimaxano replied:
`gps` prefix is redundant, right

see the comment above on `memcpy`

cmraaron reviewed (changes requested):
most of my comments are just nit picks, feel free to resolve without changes.

But the one about the function `findClosestMeasurement`, I think returns the first measurement everytime if it meets the condition, rather then the closest measurement.

It'd be great if you could add some comments in some of these math heavy functions if possible, or even tests or both.

Also it appears this branch is merged into an older version of itself.  Can you make this branch without merges?  I think just resetting to `ad4ffa840ae8fe9a629f79093e8f437dee50ff31` will do it.

dimaxano committed:
minor cleaning

dimaxano committed:
replace getRPY with getYaw

dimaxano committed:
remove explicit nanoseconds to seconds conversion

dimaxano committed:
findClosestMeasurement bugfix

dimaxano committed:
removing gps prefix

dimaxano committed:
c++ style struct definition

dimaxano committed:
Add links to original paper for some parts of the code

dimaxano commented:
@cmraaron I fixed most of the requested changes
I added code comments with the links to exact paper parts which describe the math. I think introducing initial tests to that repo is a separate non small task, so skipped that for now.

@TanmayDeshmukh @guilyx Can you please take a look again?

dimaxano commented:
@TanmayDeshmukh @guilyx
